src/v/cloud_roles/azure_aks_refresh_impl.cc

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "azure_aks_refresh_impl.h"
+
+#include "json/schema.h"
+#include "request_response_helpers.h"
+#include "utils/file_io.h"
+
+#include <seastar/coroutine/as_future.hh>
+#include <seastar/coroutine/exception.hh>
+
+#include <boost/algorithm/string/trim.hpp>
+#include <rapidjson/error/en.h>
+
+namespace {
+
+// utility to wrap std::getenv for required environment variables
+auto load_from_env(const char* env_var) -> ss::sstring {
+    auto env_value = std::getenv(env_var);
+    if (!env_value) {
+        throw std::runtime_error(fmt::format(
+          "environment variable {} is not set, the Azure AKS client cannot "
+          "function "
+          "without this.",
+          env_var));
+    }
+    return env_value;
+}
+
+constexpr static auto env_var_azure_client_id = "AZURE_CLIENT_ID";
+constexpr static auto env_var_azure_tenant_id = "AZURE_TENANT_ID";
+constexpr static auto env_var_azure_federated_token_file
+  = "AZURE_FEDERATED_TOKEN_FILE";
+constexpr static auto env_var_azure_authority_host = "AZURE_AUTHORITY_HOST";
+
+} // namespace
+
+namespace cloud_roles {
+
+azure_aks_refresh_impl::azure_aks_refresh_impl(
+  net::unresolved_address address,
+  aws_region_name region,
+  ss::abort_source& as,
+  retry_params retry_params)
+  : refresh_credentials::impl(
+    [&] {
+        if (!address.host().empty()) {
+            // non-empty host: it's an override that we should use
+            return std::move(address);
+        }
+        // empty host: use env_var
+        return net::unresolved_address{
+          load_from_env(env_var_azure_authority_host), default_port};
+    }(),
+    std::move(region),
+    as,
+    retry_params)
+  , client_id_{load_from_env(env_var_azure_client_id)}
+  , tenant_id_{load_from_env(env_var_azure_tenant_id)}
+  , federated_token_file_{load_from_env(env_var_azure_federated_token_file)} {}
+
+auto azure_aks_refresh_impl::print(std::ostream& os) const -> std::ostream& {
+    fmt::print(os, "azure_aks_refresh_impl{{address:{}}}", address());
+    return os;
+}
+
+auto azure_aks_refresh_impl::fetch_credentials() -> ss::future<api_response> {
+    // https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-client-creds-grant-flow#third-case-access-token-request-with-a-federated-credential
+    // load the encoded jwt from the token file (its content is rotated by k8s),
+    // compose:
+    // POST /{tenant_id_}/oauth2/v2.0/token HTTP/1.1               // Line
+    // breaks for clarity Host: {authority_host_}:443 Content-Type:
+    // application/x-www-form-urlencoded
```

Comment:
please fix the formatting here. you can enable/disable clang-format around this to avoid it clobbering the formatting.